gtime/hierarchical/tests/test_top_down.py

```diff
@@ -0,0 +1,178 @@
+def hierarchical_top_down_model_sga_tree_by_hand(draw, time_series_forecasting_model1_no_cache):
+    tree_adjacency_list = {'0': {}, '1': {}, '2': {'1': {}, '4': {}}, '3': {'0': {}, '2': {}}, '4': {}}
+    tree = nx.DiGraph(tree_adjacency_list)
+    root = '3'
+    return HierarchicalTopDown(model=time_series_forecasting_model1_no_cache, hierarchy_tree=tree, root=root)
+
+
+@st.composite
+def hierarchical_top_down_model_sgf_tree_by_hand(draw, time_series_forecasting_model1_no_cache):
+    tree_adjacency_list = {'0': {}, '1': {}, '2': {'1': {}, '3': {}}, '4': {'0': {}, '2': {}}, '3': {}}
+    tree = nx.DiGraph(tree_adjacency_list)
+    root = '4'
+    return HierarchicalTopDown(model=time_series_forecasting_model1_no_cache, method='tdsgf', hierarchy_tree=tree, root=root)
+
+@st.composite
+def hierarchical_top_down_model_fp_tree_by_hand(draw, time_series_forecasting_model1_no_cache):
+    tree_adjacency_list = {'0': {}, '1': {}, '2': {'1': {}, '3': {}}, '4': {'0': {}, '2': {}}, '3': {}}
+    tree = nx.DiGraph(tree_adjacency_list)
+    root = '4'
+    return HierarchicalTopDown(model=time_series_forecasting_model1_no_cache, method='tdfp', hierarchy_tree=tree, root=root)
+
+@fixture(scope="function")
+def hierarchical_basic_top_down_model(time_series_forecasting_model1_no_cache):
+    return HierarchicalTopDown(time_series_forecasting_model1_no_cache, 'infer')
+
+
+class TestHierarchicalTopDown:
+    def test_basic_constructor(self, time_series_forecasting_model1_no_cache):
+        HierarchicalTopDown(model=time_series_forecasting_model1_no_cache, hierarchy_tree='infer')
+
+    @given(dataframes=n_time_series_with_same_index())
+    def test_fit_predict_basic_top_down_on_different_data(self, dataframes, hierarchical_basic_top_down_model):
+        hierarchical_basic_top_down_model.fit(dataframes).predict(dataframes)
+
+    @given(dataframes=n_time_series_with_same_index(min_n=5))
+    def test_fit_predict_basic_top_down(self, dataframes, hierarchical_basic_top_down_model):
+        hierarchical_basic_top_down_model.fit(dataframes).predict()
+
+    @given(dataframes=n_time_series_with_same_index())
+    def test_constructor(self, time_series_forecasting_model1_no_cache, dataframes):
+        tree = tree_construction(dataframes)
+        HierarchicalTopDown(time_series_forecasting_model1_no_cache, tree)
+
+    @given(data=st.data(), dataframes=n_time_series_with_same_index(min_n=5))
+    def test_fit_predict_top_down_sga(self, data, dataframes, time_series_forecasting_model1_no_cache):
+        model = data.draw(hierarchical_top_down_model_sga(time_series_forecasting_model1_no_cache))
+        model.fit(dataframes).predict()
+
+    @given(data=st.data(), dataframes=n_time_series_with_same_index(min_n=5))
+    def test_fit_predict_top_down_sgf(self, data, dataframes, time_series_forecasting_model1_no_cache):
+        model = data.draw(hierarchical_top_down_model_sgf(time_series_forecasting_model1_no_cache))
+        model.fit(dataframes).predict()
```

Comment:
Also here, try to add at least some sanity checks.

Ideally in this case some checks on the results would be nice too, since it's not trivial as in BottomUp. Maybe if you have time you can add an ad-hoc example without hypothesis and check that the results are as expected